packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -469,18 +469,6 @@ class ConnectionStateHandler implements IConnectionStateHandler {
 				});
 			}
 			this.applyForConnectedState("addMemberEvent");
-		} else if (clientId === this.clientId) {
```

Comment:
At first I was asking the same thing you did in Teams - When would this happen?  This means that we receive the Join op for our already-connected clientId.  But we wouldn't have promoted pendingClientId to clientId without seeing this Join op, at least that's the mainstream assumption.

But there's a caveat - we set clientId from pendingLocalState when rehydrating.  If the snapshot was before the join op, the Container will replay that join op (the serialized state includes all sequenced ops following the snapshot until time of serialization, they're replayed in Container.load), but the clientId will already be set.

Once we see that join op for the stashed clientId, we know that any subsequent local ops we rehydrated with are a "liability" - maybe they were sequenced with this clientId already.  So until we see that corresponding leave op, we can't enter "connected" state on a new clientId.

So this seems correct to me.  That said, I couldn't quite understand the tests that are covering this, so I'm not 100% sure I got it right.  Also, I would expect a similar affordance for the case where the serialized clientId's join op is included in the snapshot, but I'm not sure if/where that's covered.

Reply:
here, clientId !== pendingClientId. This means this.clientId is for old connection.
I'd need to better understand pendingLocalState flow (where maybe this logic makes sense), but it does not make sense for "regular" case - we already made all the right decision about what to do about old connection by the point we established new pendingClientId (or maybe it's undefined, which is not much different). This code should not make similar decisions but use different rules - that would never work

Reply:
Ah, I see where the problem with pendingLocalState. It bends time. It stores clientId from the future (relative to what this class will observe as it plays ops)! This is not great, patching this layer to support this scenario would not be great - it's a leakage that (ideally) we want to avoid, if that's possible (i.e. ideally this layer should not know and should not care about such scenario, it should ideally be implemented outside of this layer). Now is the hard part - having some proposal :)